File: src/buddy_cli/runtime_manager.py
"""Discover, install, and start Ollama runtimes."""

from __future__ import annotations

import os
import shutil
import stat
import subprocess
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Optional
from urllib import parse

from buddy_cli.archive import extract_archive
from buddy_cli.config import BuddyConfig
from buddy_cli.constants import (
    DEFAULT_OLLAMA_BASE_URL,
    MANAGED_OLLAMA_BASE_URL,
    OLLAMA_RUNTIME_VERSION,
)
from buddy_cli.download import DownloadProgress, download_verified, sha256_file
from buddy_cli.ollama import OllamaClient, OllamaError
from buddy_cli.paths import AppPaths
from buddy_cli.runtime_manifest import RuntimeSpec, resolve_runtime_spec


class RuntimeManagerError(RuntimeError):
    """Raised when an Ollama runtime cannot be prepared."""


@dataclass(frozen=True)
class RuntimeSelection:
    provider: str
    base_url: str
    executable: Optional[Path]
    runtime_version: Optional[str]


RuntimeStatus = Callable[[str], None]


class RuntimeManager:
    """Manage system and Buddy-owned Ollama processes."""

    def __init__(self, paths: AppPaths) -> None:
        self.paths = paths

    @staticmethod
    def api_is_ready(base_url: str) -> bool:
        try:
            OllamaClient(base_url, timeout=1.0).get_version()
            return True
        except (OllamaError, ValueError):
            return False

    @staticmethod
    def find_system_executable() -> Optional[Path]:
        executable = shutil.which("ollama")
        if executable:
            return Path(executable)

        candidates = []
        if sys.platform == "darwin":
            candidates.extend(
                [
                    Path("/Applications/Ollama.app/Contents/Resources/ollama"),
                    Path.home()
                    / "Applications"
                    / "Ollama.app"
                    / "Contents"
                    / "Resources"
                    / "ollama",
                ]
            )
        elif sys.platform == "win32":
            local_app_data = os.environ.get("LOCALAPPDATA")
            if local_app_data:
                candidates.append(
                    Path(local_app_data) / "Programs" / "Ollama" / "ollama.exe"
                )

        return next((path for path in candidates if path.is_file()), None)

    def installation_dir(self, spec: RuntimeSpec) -> Path:
        return self.paths.runtimes_dir / spec.installation_name

    @staticmethod
    def _find_executable(root: Path) -> Optional[Path]:
        candidates = ("ollama.exe",) if os.name == "nt" else ("ollama",)
        for name in candidates:
            matches = [path for path in root.rglob(name) if path.is_file()]
            if matches:
                matches.sort(key=lambda path: (len(path.parts), str(path)))
                return matches[0]
        return None

    def find_managed_executable(self) -> Optional[Path]:
        try:
            spec = resolve_runtime_spec()
        except Exception:
            return None
        installation = self.installation_dir(spec)
        return self._find_executable(installation) if installation.exists() else None

    def install_managed(
        self,
        *,
        download_progress: Optional[DownloadProgress] = None,
        status_callback: Optional[RuntimeStatus] = None,
    ) -> RuntimeSelection:
        spec = resolve_runtime_spec()
        self.paths.ensure_directories()
        installation = self.installation_dir(spec)
        existing = (
            self._find_executable(installation) if installation.exists() else None
        )
        if existing:
            return RuntimeSelection(
                provider="managed",
                base_url=MANAGED_OLLAMA_BASE_URL,
                executable=existing,
                runtime_version=spec.version,
            )

        archive = self.paths.downloads_dir / spec.asset_name
        if archive.exists() and sha256_file(archive) != spec.sha256:
            archive.unlink()
        if not archive.exists():
            if status_callback:
                status_callback(f"Downloading Ollama {spec.version}")
            download_verified(
                spec.url,
                archive,
                spec.sha256,
                progress=download_progress,
            )

        staged: Optional[Path] = Path(
            tempfile.mkdtemp(prefix=".ollama-stage-", dir=str(self.paths.runtimes_dir))
        )
        try:
            if status_callback:
                status_callback("Extracting the verified Ollama runtime")
            extract_archive(archive, staged, spec.archive_type)
            staged_executable = self._find_executable(staged)
            if staged_executable is None:
                raise RuntimeManagerError(
                    "the Ollama archive did not contain an executable"
                )

            if os.name != "nt":
                staged_executable.chmod(
                    staged_executable.stat().st_mode | stat.S_IXUSR | stat.S_IXGRP
                )
            relative_executable = staged_executable.relative_to(staged)

            if installation.exists():
                shutil.rmtree(installation)
            os.replace(str(staged), str(installation))
            staged = None
            archive.unlink(missing_ok=True)
            executable = installation / relative_executable
            return RuntimeSelection(
                provider="managed",
                base_url=MANAGED_OLLAMA_BASE_URL,
                executable=executable,
                runtime_version=spec.version,
            )
        finally:
            if staged is not None and staged.exists():
                shutil.rmtree(staged)

    def start(self, selection: RuntimeSelection) -> None:
        if self.api_is_ready(selection.base_url):
            return
        if selection.executable is None or not selection.executable.exists():
            raise RuntimeManagerError("the configured Ollama executable does not exist")

        self.paths.ensure_directories()
        environment = os.environ.copy()
        if selection.provider == "managed":
            parsed = parse.urlparse(selection.base_url)
            host = parsed.hostname or "127.0.0.1"
            port = parsed.port or 11435
            environment["OLLAMA_HOST"] = f"{host}:{port}"
            environment["OLLAMA_MODELS"] = str(self.paths.models_dir)

        creation_flags = 0
        popen_options = {"start_new_session": True}
        if os.name == "nt":
            creation_flags = getattr(
                subprocess, "CREATE_NEW_PROCESS_GROUP", 0
            ) | getattr(subprocess, "DETACHED_PROCESS", 0)
            popen_options = {}

        with self.paths.ollama_log_file.open("ab") as log_handle:
            try:
                process = subprocess.Popen(
                    [str(selection.executable), "serve"],
                    stdin=subprocess.DEVNULL,
                    stdout=log_handle,
                    stderr=subprocess.STDOUT,
                    env=environment,
                    creationflags=creation_flags,
                    **popen_options,
                )
            except OSError as exc:
                raise RuntimeManagerError(f"could not start Ollama: {exc}") from exc

        self.paths.ollama_pid_file.write_text(f"{process.pid}\n", encoding="utf-8")
        try:
            OllamaClient(selection.base_url, timeout=1.0).wait_until_ready(timeout=45)
        except OllamaError:
            process.terminate()
            raise

    def selection_from_config(self, config: BuddyConfig) -> RuntimeSelection:
        executable = Path(config.executable) if config.executable else None
        return RuntimeSelection(
            provider=config.provider,
            base_url=config.base_url,
            executable=executable,
            runtime_version=config.runtime_version,
        )

    def discover(
        self, config: Optional[BuddyConfig] = None
    ) -> Optional[RuntimeSelection]:
        if config:
            configured = self.selection_from_config(config)
            if self.api_is_ready(configured.base_url):
                return configured
            if configured.executable and configured.executable.exists():
                return configured

        if self.api_is_ready(DEFAULT_OLLAMA_BASE_URL):
            return RuntimeSelection(
                provider="system",
                base_url=DEFAULT_OLLAMA_BASE_URL,
                executable=self.find_system_executable(),
                runtime_version=None,
            )

        system_executable = self.find_system_executable()
        if system_executable:
            return RuntimeSelection(
                provider="system",
                base_url=DEFAULT_OLLAMA_BASE_URL,
                executable=system_executable,
                runtime_version=None,
            )

        managed_executable = self.find_managed_executable()
        if managed_executable:
            return RuntimeSelection(
                provider="managed",
                base_url=MANAGED_OLLAMA_BASE_URL,
                executable=managed_executable,
                runtime_version=OLLAMA_RUNTIME_VERSION,
            )
        return None

File: src/runtime_manager.rs
//! Discover, install, and start Ollama runtimes.

use std::env;
use std::fmt;
use std::fs::{self, OpenOptions};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::Duration;
use tempfile::Builder;

use crate::archive::extract_archive;
use crate::config::BuddyConfig;
use crate::constants::{DEFAULT_OLLAMA_BASE_URL, MANAGED_OLLAMA_BASE_URL, OLLAMA_RUNTIME_VERSION};
use crate::download::{
    download_verified, sha256_file, DownloadOptions, DownloadProgress, DownloadStatus,
};
use crate::ollama::OllamaClient;
use crate::paths::AppPaths;
use crate::runtime_manifest::{resolve_runtime_spec, RuntimeSpec};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuntimeManagerError(pub String);

impl fmt::Display for RuntimeManagerError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for RuntimeManagerError {}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuntimeSelection {
    pub provider: String,
    pub base_url: String,
    pub executable: Option<PathBuf>,
    pub runtime_version: Option<String>,
}

pub struct RuntimeManager {
    pub paths: AppPaths,
}

impl RuntimeManager {
    pub fn new(paths: AppPaths) -> Self {
        Self { paths }
    }

    pub fn api_is_ready(base_url: &str) -> bool {
        if let Ok(client) = OllamaClient::with_timeout(base_url, Duration::from_secs(1)) {
            client.get_version().is_ok()
        } else {
            false
        }
    }

    pub fn find_system_executable() -> Option<PathBuf> {
        if let Some(path_var) = env::var_os("PATH") {
            for dir in env::split_paths(&path_var) {
                let bin = dir.join(if cfg!(windows) {
                    "ollama.exe"
                } else {
                    "ollama"
                });
                if bin.is_file() {
                    return Some(bin);
                }
            }
        }

        let mut candidates = Vec::new();
        if cfg!(target_os = "macos") {
            candidates.push(PathBuf::from(
                "/Applications/Ollama.app/Contents/Resources/ollama",
            ));
            if let Some(home) = dirs::home_dir() {
                candidates.push(home.join("Applications/Ollama.app/Contents/Resources/ollama"));
            }
        } else if cfg!(target_os = "windows") {
            if let Ok(local_app_data) = env::var("LOCALAPPDATA") {
                candidates.push(PathBuf::from(local_app_data).join("Programs/Ollama/ollama.exe"));
            }
        }

        candidates.into_iter().find(|path| path.is_file())
    }

    pub fn installation_dir(&self, spec: &RuntimeSpec) -> PathBuf {
        self.paths.runtimes_dir().join(spec.installation_name())
    }

    fn find_executable_in_dir(root: &Path) -> Option<PathBuf> {
        let target_name = if cfg!(windows) {
            "ollama.exe"
        } else {
            "ollama"
        };
        let mut matches = Vec::new();

        fn walk(dir: &Path, target: &str, matches: &mut Vec<PathBuf>) {
            if let Ok(entries) = fs::read_dir(dir) {
                for entry in entries.flatten() {
                    let path = entry.path();
                    if path.is_dir() {
                        walk(&path, target, matches);
                    } else if path.file_name().and_then(|n| n.to_str()) == Some(target) {
                        matches.push(path);
                    }
                }
            }
        }

        walk(root, target_name, &mut matches);
        if matches.is_empty() {
            None
        } else {
            matches.sort_by_key(|p| (p.components().count(), p.to_string_lossy().to_string()));
            Some(matches.remove(0))
        }
    }

    pub fn find_managed_executable(&self) -> Option<PathBuf> {
        let spec = resolve_runtime_spec().ok()?;
        let installation = self.installation_dir(&spec);
        if installation.exists() {
            Self::find_executable_in_dir(&installation)
        } else {
            None
        }
    }

    pub fn install_managed<'a>(
        &self,
        download_progress: Option<DownloadProgress<'a>>,
        mut status_callback: Option<DownloadStatus<'a>>,
    ) -> Result<RuntimeSelection, Box<dyn std::error::Error>> {
        let spec = resolve_runtime_spec().map_err(|e| RuntimeManagerError(e.to_string()))?;
        self.paths.ensure_directories()?;

        let installation = self.installation_dir(&spec);
        if installation.exists() {
            if let Some(existing) = Self::find_executable_in_dir(&installation) {
                return Ok(RuntimeSelection {
                    provider: "managed".to_string(),
                    base_url: MANAGED_OLLAMA_BASE_URL.to_string(),
                    executable: Some(existing),
                    runtime_version: Some(spec.version.to_string()),
                });
            }
        }

        let archive = self.paths.downloads_dir().join(spec.asset_name);
        if archive.exists() {
            if let Ok(current_sha) = sha256_file(&archive) {
                if !current_sha.eq_ignore_ascii_case(spec.sha256) {
                    let _ = fs::remove_file(&archive);
                }
            }
        }

        if !archive.exists() {
            if let Some(ref mut s) = status_callback {
                s(&format!("Downloading Ollama {}", spec.version));
            }

            let download_opts = DownloadOptions {
                progress: download_progress,
                status: status_callback,
                ..Default::default()
            };

            download_verified(&spec.url(), &archive, spec.sha256, download_opts)?;
        }

        let stage_temp = Builder::new()
            .prefix(".ollama-stage-")
            .tempdir_in(self.paths.runtimes_dir())?;
        let staged_path = stage_temp.path().to_path_buf();

        extract_archive(&archive, &staged_path, spec.archive_type)?;

        let staged_executable = Self::find_executable_in_dir(&staged_path).ok_or_else(|| {
            RuntimeManagerError("the Ollama archive did not contain an executable".to_string())
        })?;

        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            if let Ok(meta) = fs::metadata(&staged_executable) {
                let mut perms = meta.permissions();
                perms.set_mode(perms.mode() | 0o755);
                let _ = fs::set_permissions(&staged_executable, perms);
            }
        }

        let relative_executable = staged_executable
            .strip_prefix(&staged_path)
            .map_err(|e| RuntimeManagerError(e.to_string()))?
            .to_path_buf();

        if installation.exists() {
            let _ = fs::remove_dir_all(&installation);
        }

        let raw_stage_path = stage_temp.keep();
        fs::rename(&raw_stage_path, &installation)?;
        let _ = fs::remove_file(&archive);

        let final_executable = installation.join(relative_executable);

        Ok(RuntimeSelection {
            provider: "managed".to_string(),
            base_url: MANAGED_OLLAMA_BASE_URL.to_string(),
            executable: Some(final_executable),
            runtime_version: Some(spec.version.to_string()),
        })
    }

    pub fn start(&self, selection: &RuntimeSelection) -> Result<(), Box<dyn std::error::Error>> {
        if Self::api_is_ready(&selection.base_url) {
            return Ok(());
        }

        let executable = selection.executable.as_ref().ok_or_else(|| {
            RuntimeManagerError("the configured Ollama executable does not exist".to_string())
        })?;

        if !executable.exists() {
            return Err(Box::new(RuntimeManagerError(
                "the configured Ollama executable does not exist".to_string(),
            )));
        }

        self.paths.ensure_directories()?;

        let mut cmd = Command::new(executable);
        cmd.arg("serve");
        cmd.stdin(Stdio::null());

        let log_file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(self.paths.ollama_log_file())?;
        cmd.stdout(log_file.try_clone()?);
        cmd.stderr(log_file);

        if selection.provider == "managed" {
            let port = if selection.base_url.contains("11435") {
                "11435"
            } else {
                "11434"
            };
            cmd.env("OLLAMA_HOST", format!("127.0.0.1:{}", port));
            cmd.env("OLLAMA_MODELS", self.paths.models_dir());
        }

        #[cfg(unix)]
        {
            use std::os::unix::process::CommandExt;
            unsafe {
                cmd.pre_exec(|| {
                    libc::setsid();
                    Ok(())
                });
            }
        }

        #[cfg(windows)]
        {
            use std::os::windows::process::CommandExt;
            const CREATE_NEW_PROCESS_GROUP: u32 = 0x00000200;
            const DETACHED_PROCESS: u32 = 0x00000008;
            cmd.creation_flags(CREATE_NEW_PROCESS_GROUP | DETACHED_PROCESS);
        }

        let child = cmd
            .spawn()
            .map_err(|e| RuntimeManagerError(format!("could not start Ollama: {}", e)))?;

        let pid = child.id();
        let _ = fs::write(self.paths.ollama_pid_file(), format!("{}\n", pid));

        let client = OllamaClient::with_timeout(&selection.base_url, Duration::from_secs(1))
            .map_err(|e| RuntimeManagerError(e.to_string()))?;

        if let Err(e) = client.wait_until_ready(Duration::from_secs(45), Duration::from_millis(250))
        {
            return Err(Box::new(RuntimeManagerError(format!(
                "Ollama did not become ready: {}",
                e
            ))));
        }

        Ok(())
    }

    pub fn selection_from_config(&self, config: &BuddyConfig) -> RuntimeSelection {
        RuntimeSelection {
            provider: config.provider.clone(),
            base_url: config.base_url.clone(),
            executable: config.executable.as_ref().map(PathBuf::from),
            runtime_version: config.runtime_version.clone(),
        }
    }

    pub fn discover(&self, config: Option<&BuddyConfig>) -> Option<RuntimeSelection> {
        if let Some(cfg) = config {
            let configured = self.selection_from_config(cfg);
            if Self::api_is_ready(&configured.base_url) {
                return Some(configured);
            }
            if let Some(ref exe) = configured.executable {
                if exe.exists() {
                    return Some(configured);
                }
            }
        }

        if Self::api_is_ready(DEFAULT_OLLAMA_BASE_URL) {
            return Some(RuntimeSelection {
                provider: "system".to_string(),
                base_url: DEFAULT_OLLAMA_BASE_URL.to_string(),
                executable: Self::find_system_executable(),
                runtime_version: None,
            });
        }

        if let Some(sys_exe) = Self::find_system_executable() {
            return Some(RuntimeSelection {
                provider: "system".to_string(),
                base_url: DEFAULT_OLLAMA_BASE_URL.to_string(),
                executable: Some(sys_exe),
                runtime_version: None,
            });
        }

        if let Some(managed_exe) = self.find_managed_executable() {
            return Some(RuntimeSelection {
                provider: "managed".to_string(),
                base_url: MANAGED_OLLAMA_BASE_URL.to_string(),
                executable: Some(managed_exe),
                runtime_version: Some(OLLAMA_RUNTIME_VERSION.to_string()),
            });
        }

        None
    }
}
